cpp/include/raft/comms/std_comms.hpp

```
@@ -0,0 +1,406 @@
+   * @param comm initilized nccl communicator
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, int size, int rank)
+    : _nccl_comm(comm), _size(size), _rank(rank) {
+    initialize();
+  };
+
+  virtual ~std_comms() {
+    CUDA_CHECK_NO_THROW(cudaStreamDestroy(_stream));
+
+    CUDA_CHECK_NO_THROW(cudaFree(_sendbuff));
+    CUDA_CHECK_NO_THROW(cudaFree(_recvbuff));
+  }
+
+  void initialize() {
+    CUDA_CHECK(cudaStreamCreate(&_stream));
+
+    CUDA_CHECK(cudaMalloc(&_sendbuff, sizeof(int)));
+    CUDA_CHECK(cudaMalloc(&_recvbuff, sizeof(int)));
+  }
+
+  int getSize() const { return _size; }
+
+  int getRank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> commSplit(int color, int key) const {
+    // Not supported by NCCL
+    ASSERT(false,
+           "ERROR: commSplit called but not yet supported in this comms "
+           "implementation.");
+  }
+
+  void barrier() const {
+    CUDA_CHECK(cudaMemsetAsync(_sendbuff, 1, sizeof(int), _stream));
+    CUDA_CHECK(cudaMemsetAsync(_recvbuff, 1, sizeof(int), _stream));
+
+    allreduce(_sendbuff, _recvbuff, 1, INT, SUM, _stream);
+
+    ASSERT(syncStream(_stream) == status_t::commStatusSuccess,
+           "ERROR: syncStream failed. This can be caused by a failed rank.");
+  }
+
+  void get_request_id(request_t *req) const {
+    request_t req_id;
+
+    if (this->_free_requests.empty())
+      req_id = this->_next_request_id++;
+    else {
+      auto it = this->_free_requests.begin();
+      req_id = *it;
+      this->_free_requests.erase(it);
+    }
+    *req = req_id;
+  }
+
+  void isend(const void *buf, int size, int dest, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    get_request_id(request);
+    ucp_ep_h ep_ptr = (*_ucp_eps)[dest];
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
```

What  happens if malloc fails here?

---

No call has been made to the UCX API so the sender would fail and any receiving ranks would timeout.

---

If I understand this correctly, if malloc fails, it is better to throw an exception than continue execution.

I think
```
auto ucp_request = std::make_unique<ucp_request>();
```
will be safer.